src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java

```diff
@@ -1,23 +1,742 @@
+    if (shippingInstructionTO.getCarrierBookingReference() == null) {
+      List<CargoItemTO> cargoItems = new ArrayList<>();
+      for (ShipmentEquipmentTO shipmentEquipmentTO :
+          shippingInstructionTO.getShipmentEquipments()) {
+        cargoItems.addAll(shipmentEquipmentTO.getCargoItems());
+      }
+      return cargoItems.get(0).getCarrierBookingReference();
     }
+    return shippingInstructionTO.getCarrierBookingReference();
+  }
+
+  private Mono<ShipmentEvent> createShipmentEvent(ShippingInstruction shippingInstruction) {
+    return shipmentEventFromShippingInstruction(shippingInstruction, null)
+        .flatMap(shipmentEventService::create)
+        .switchIfEmpty(
+            Mono.error(
+                ConcreteRequestErrorMessageException.invalidParameter(
+                    "Failed to create shipment event for ShippingInstruction.")));
+  }
+
+  private Mono<ShipmentEvent> shipmentEventFromShippingInstruction(
+      ShippingInstruction shippingInstruction, String reason) {
+    ShipmentEvent shipmentEvent = new ShipmentEvent();
+    shipmentEvent.setShipmentEventTypeCode(
+        ShipmentEventTypeCode.valueOf(shippingInstruction.getDocumentStatus().name()));
+    shipmentEvent.setDocumentTypeCode(DocumentTypeCode.SHI);
+    shipmentEvent.setEventClassifierCode(EventClassifierCode.ACT);
+    shipmentEvent.setEventType(null);
+    shipmentEvent.setCarrierBookingReference(null);
+    shipmentEvent.setDocumentID(shippingInstruction.getShippingInstructionID());
+    shipmentEvent.setEventCreatedDateTime(OffsetDateTime.now());
+    shipmentEvent.setEventDateTime(shippingInstruction.getShippingInstructionUpdatedDateTime());
+    shipmentEvent.setReason(reason);
+    return Mono.just(shipmentEvent);
+  }
+
+  @Override
+  public Mono<ShippingInstructionTO> replaceOriginal(
+      String shippingInstructionID, ShippingInstructionTO update) {
+    return null;
+  }
+
+  //    private Mono<Void> processFreightPayableAt(ShippingInstructionTO shippingInstructionTO,
+  // ShippingInstruction shippingInstruction) {
+  //        return Mono.justOrEmpty(shippingInstructionTO.getPlaceOfIssue())
+  //                .flatMap(locationService::ensureResolvable)
+  //                .doOnNext(shippingInstructionTO::setPlaceOfIssue)
+  //                .map(LocationTO::getId)
+  //                .doOnNext(shippingInstruction::setFreightPayableAt)
+  //                .then();
+  //    }
 
+  //    private Mono<ShippingInstructionTO> extractShipmentRelatedFields(ShippingInstructionTO
+  // shippingInstructionTO,
+  //                                                                     List<UUID> shipmentIDs,
+  //                                                                     List<Tuple2<CargoItem,
+  // CargoItemTO>> cargoItemTuples
+  //    ) {
+  //        Map<UUID, List<CargoItemTO>> shipmentEquipmentID2CargoItems =
+  // cargoItemTuples.stream().collect(
+  //                Collectors.groupingBy(
```

Comment:
Please remove this code if it is no longer needed...

Reply:
Done